feat: add DDSketch quantile

srikanthccv opened the pull request:
Fixes https://github.com/ClickHouse/ClickHouse/issues/6739

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `quantileDDSketch` aggregate function as well as the corresponding `quantilesDDSketch` and `medianDDSketch`. It is based on the DDSketch https://www.vldb.org/pvldb/vol12/p2195-masson.pdf.
### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)


- Motivation: It is very important to calculate accurate quantiles fast. The OpenTelemetry project is the emerging standard in the observability ecosystem, and the preferred way to record the latencies is [`ExponentialHistogram`](https://github.com/open-telemetry/opentelemetry-proto/blob/cfbf9357c03bf4ac150a3ab3bcbe4cc4ed087362/opentelemetry/proto/metrics/v1/metrics.proto#L222). Having this feature in ClickHouse makes it more adaptable.   
- Parameters: relative accuracy 
- Example use: 

```sql
SELECT quantileDDSketch(0.01, 0.42)(number)
FROM numbers_mt(1000000000)

Query id: 0f273858-44eb-4359-9108-198f0179fc5e

┌─quantileDDSketch(0.01, 0.42)(number)─┐
│                   417841070.41428167 │
└──────────────────────────────────────┘

```

```sql
CREATE TABLE default.comparision
(
    `id` UInt64,
    `sketch` AggregateFunction(quantilesDDSketch(0.02, 0.9), UInt64),
    `td` AggregateFunction(quantilesTDigest(0.5, 0.9, 0.99), UInt64),
    `bf` AggregateFunction(quantilesBFloat16(0.5, 0.9, 0.99), UInt64)
)
ENGINE = SummingMergeTree
ORDER BY id
SETTINGS index_granularity = 8192


insert into comparision
select
    number % 10000 as id,
    quantilesDDSketchState(0.02, 0.9)(number) as sketch,
    quantilesTDigestState(0.5, 0.9, 0.99)(number) as td,
    quantilesBFloat16State(0.5, 0.9, 0.99)(number) as bf  
from numbers(100000000)
group by id;


SELECT
    quantilesDDSketchMerge(0.02, 0.9)(sketch) AS sketch
FROM comparision

SELECT
    quantileTDigestMerge(0.9)(td) AS td
FROM comparision

SELECT
    quantileBFloat16Merge(0.9)(bf) AS bf
FROM comparision

```



> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

srikanthccv committed:
feat: add DDSketch quantile

srikanthccv commented:
@SmitaRKulkarni I had to close this https://github.com/ClickHouse/ClickHouse/pull/55305 because of a messed up rebase. Please review this instead. I added the docs and address comments from the other PR

srikanthccv commented:
I would appreciate it if someone could trigger workflows.

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #56342 from srikanthccv/ClickHouse:ddsketch-quantile -->
*This is an automated comment for commit 0c660e8c15c3534883acdba9acfe37dbf2b84df7 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56342/0c660e8c15c3534883acdba9acfe37dbf2b84df7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6762929750">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56342/0c660e8c15c3534883acdba9acfe37dbf2b84df7/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

srikanthccv committed:
Fix StyleCheck

srikanthccv committed:
Move quantileddsketch.md under right dir

srikanthccv committed:
Update stateless test number and add stateful test

srikanthccv commented:
How can I get the fuzzer to pass a reasonable relative accuracy? The value `1.1754943508222875e-38` (understandably) crashes the build. I am thinking of throwing an exception in case of provided relative accuracy is too low. I think most practical applications will be fine with something like `0.0003` i.e. no more than 0.03%. Let me know what you think.

bharatnc commented:
> How can I get the fuzzer to pass a reasonable relative accuracy? The value `1.1754943508222875e-38` (understandably) crashes the build. I am thinking of throwing an exception in case of provided relative accuracy is too low. I think most practical applications will be fine with something like `0.0003` i.e. no more than 0.03%. Let me know what you think.

It looks like the msan fuzzer builds indicate possible memory violations here: [src/AggregateFunctions/QuantileDDSketch.h:25](https://github.com/ClickHouse/ClickHouse/pull/56342/files#diff-46c443dd726a9de237c408e3a82affda8185adc50c54cc5ea325184304cc962bR25) and here [src/AggregateFunctions/QuantileDDSketch.h:31](https://github.com/ClickHouse/ClickHouse/pull/56342/files#diff-46c443dd726a9de237c408e3a82affda8185adc50c54cc5ea325184304cc962bR31). 

You could alternatively handle arithmetic operations in a way such that they handle the overflows correctly ? There are usages in other places which leverage these [arithmeticOverflow](https://github.com/ClickHouse/ClickHouse/blob/master/base/base/arithmeticOverflow.h) functions that handle overflows for you.

srikanthccv commented:
>SELECT round(quantileDDSketch(1.1754943508222875e-38)(number)) FROM numbers(255) GROUP BY -2 / (number + 3) WITH ROLLUP

They are failing because the passed relative accuracy is too low which demands bins beyond allocatable and program crashes. If you change the relative accuracy to a reasonable value it works without crashing.

srikanthccv committed:
Throw exception when relative accuracy is too low

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Update test number

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
QuantileBFloat16 test is failing for [ubsan, ParallelReplicas](https://s3.amazonaws.com/clickhouse-test-reports/56342/5b593fb482244c28dd2013e5fb503268c573bbdb/stateful_tests__ubsan__parallelreplicas_.html) and QuantileDDSketch if failing for [ubsan](https://s3.amazonaws.com/clickhouse-test-reports/56342/5b593fb482244c28dd2013e5fb503268c573bbdb/stateful_tests__ubsan_.html). It seems like a timeout. How can I troubleshoot or are they occasional intermittent failures?

srikanthccv commented:
I will resolve the conflicts

bharatnc commented:
> QuantileBFloat16 test is failing for [ubsan, ParallelReplicas](https://s3.amazonaws.com/clickhouse-test-reports/56342/5b593fb482244c28dd2013e5fb503268c573bbdb/stateful_tests__ubsan__parallelreplicas_.html) and QuantileDDSketch if failing for [ubsan](https://s3.amazonaws.com/clickhouse-test-reports/56342/5b593fb482244c28dd2013e5fb503268c573bbdb/stateful_tests__ubsan_.html). It seems like a timeout. How can I troubleshoot or are they occasional intermittent failures?

I think QuantileBFloat16 test failure is un-related, it's failing because its not able to connect to the server (likely the server has crashed from the other test failure.)

The main error seems to be the following

```
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /build/src/AggregateFunctions/Store.h:106:35 in 
```
Possible overflows are happening here during addition. Probably, the easiest way is to fix is to use `NO_SANITIZE_UNDEFINED` (i.e. to suppress undefined behavior sanitizer) in the function signature of serialize. Here is an example for [this](https://github.com/ClickHouse/ClickHouse/blob/master/src/AggregateFunctions/AggregateFunctionAvg.h#L45). There are also [arithmeticOverflow](https://github.com/ClickHouse/ClickHouse/blob/master/base/base/arithmeticOverflow.h) helper functions that basically do the same (suppresses the UBSAN) to handle the overflows for you while doing arithmetic operations. Better way is to actually handle overflows without suppressing the UBSAN if you're able to do so realistically .

srikanthccv committed:
Resolve merge conflicts

srikanthccv committed:
Fix undefined behaviour for empty store

srikanthccv commented:
@bharatnc Thanks, that was helpful. I saw other files but didn't pay attention to `stderr.log`. This happens when the store is empty. I pushed a fix (check for the empty state) to solve it.

srikanthccv committed:
Fix quantileGK test

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Update test numbers

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
All the main tests are passing now; failures are unrelated to this change. @SmitaRKulkarni, please review.

bharatnc commented on src/AggregateFunctions/Mapping.h:
```diff
@@ -0,0 +1,112 @@
+        return min_possible;
+    }
+
+    Float64 getMaxPossible() const
```
It doesn't seem like this function is actually used anywhere.

srikanthccv replied:
I can remove this

bharatnc commented on src/AggregateFunctions/AggregateFunctionQuantile.h:
```diff
@@ -87,7 +91,36 @@ class AggregateFunctionQuantile final
+                    ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Aggregate function {} requires at least one param", getName());
+
+            const auto & relative_accuracy_field = params[0];
+            if (!isFloat64FieldType(relative_accuracy_field.getType()))
```
I think you can actually use `argument_types_` to find out if the param is a float64 like so:
```
            if(!WhichDataType(argument_types_[0]).isFloat64())
```
If that works, you possibly won't need to add this new function `isFloat64FieldType()`.

srikanthccv replied:
We want to get value and type both and I saw `isInt64OrUInt64FieldType` being used elsewhere in the codebase and used it for simplicity. I don't see any advantage of this over it.

bharatnc commented on src/AggregateFunctions/DDSketchEncoding.h:
```diff
@@ -0,0 +1,42 @@
+        UInt8 byte;
+        Flag(UInt8 t, UInt8 s) : byte(t | s) { }
+        UInt8 Type() const { return byte & flagTypeMask; }
+        UInt8 SubFlag() const { return byte & subFlagMask; }
```
This function doesn't seem to be used anywhere. You could remove it or perhaps do:
```suggestion
        [[maybe_unused]] UInt8 SubFlag() const { return byte & subFlagMask; }
```

bharatnc commented on src/AggregateFunctions/DDSketchEncoding.h:
```diff
@@ -0,0 +1,42 @@
+
+    // BINS
+    static constexpr UInt8 BinEncodingIndexDeltasAndCounts = 1 << numBitsForType;
+    static constexpr UInt8 BinEncodingIndexDeltas = 2 << numBitsForType;
```
Similarly, this doesn't seem to be used anywhere as well.

srikanthccv replied:
There are indeed some unused flags in the encoding.h but I left them deliberately to make the encoding consistent with the original sketch and be able to implement alternative store or mapping implementation in future.

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Update src/AggregateFunctions/

srikanthccv commented:
@alexey-milovidov can you review this? It would be really great if we could get this reviewed and merged.

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Throw exception on out of range values

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

al13n321 commented on src/AggregateFunctions/Mapping.h:
```diff
@@ -0,0 +1,116 @@
+    explicit LogarithmicMapping(Float64 relative_accuracy_, Float64 offset_ = 0.0)
+        : KeyMapping(relative_accuracy_, offset_)
+    {
+        multiplier *= std::log(2);
```
What's the purpose of multiplying by log(2), then dividing by it again in a way that exactly cancels out? Wouldn't it be equivalent if you remove both log(2)s and replace pow(2.0) with exp()?

al13n321 commented on src/AggregateFunctions/Mapping.h:
```diff
@@ -0,0 +1,116 @@
+    Float64 min_possible;
+    Float64 max_possible;
+    Float64 multiplier;
+    Float64 offset;
```
Curious: What's the purpose of adding offset to the key? I can imagine offset being useful for *values*, e.g. add 1 to avoid log(0) or subtract 2025-01-01 from dates to increase resolution where it matters.

al13n321 commented on src/AggregateFunctions/DDSketch.h:
```diff
@@ -0,0 +1,239 @@
+    Float64 zero_count;
+    Float64 count;
+    DDSketchEncoding enc;
+    Poco::Logger * log = &Poco::Logger::get("DDSketch");
```
Unused? Also, doesn't belong here, better to create it when something needs to be logger rather than for every row of aggregation results - the latter can be much more frequent (and the former is amortized by the cost of logging a message).

al13n321 reviewed (changes requested):
Looks correct. Many review comments, the only important ones are:
 * Cover more corner cases in the test.
 * Check that it's not slow (compared to e.g. other quantile aggregate functions).

al13n321 commented on docs/en/sql-reference/aggregate-functions/reference/quantileddsketch.md:
```diff
@@ -0,0 +1,61 @@
+
+**Parameters**
+
+- `relative_accuracy` — Relative accuracy of the quantile. Possible values are in the range from 0 to 1. [Float](../../../sql-reference/data-types/float.md).
```
Would it be better to make it optional, with default value e.g. 0.01?

srikanthccv commented:
Thanks for the review @al13n321. I will add more test cases and address the review comments. 

>Check that it's not slow (compared to e.g. other quantile aggregate functions)

I tested again the quantile tdigest and quantile bfloat16 and it's faster up to 2-3x for the 1% relative accuracy.

srikanthccv committed:
Update relative accuracy docs and add encoding details

srikanthccv committed:
Address review comments and suggestions

srikanthccv committed:
Remove unused function

srikanthccv commented on tests/queries/0_stateless/02919_ddsketch_quantile.sql:
```diff
@@ -0,0 +1,5 @@
+SELECT round(quantileDDSketch(0.01, 0.5)(number), 2) FROM numbers(200);
+SELECT round(quantileDDSketch(0.01, 0.69)(number), 2) FROM numbers(500);
+SELECT round(quantileDDSketch(0.01, 0.42)(number), 2) FROM numbers(200);
+SELECT round(quantileDDSketch(0.01, 0.99)(number), 2) FROM numbers(500);
```
Add more test cases

srikanthccv commented on src/AggregateFunctions/Store.h:
```diff
@@ -0,0 +1,289 @@
+                previous_index = index;
+            }
+        }
+        sparse_encoding_size += estimatedVarIntSize(num_non_empty_bins);
```
Done

alexey-milovidov commented on src/Core/Field.h:
```diff
@@ -854,6 +854,11 @@ inline constexpr bool isInt64OrUInt64orBoolFieldType(Field::Types::Which t)
         || t == Field::Types::Bool;
 }
 
+inline constexpr bool isFloat64FieldType(Field::Types::Which t)
```
Doesn't look needed.

alexey-milovidov commented on tests/queries/0_stateless/temp:
```diff
@@ -0,0 +1,25 @@
+localhost :) SELECT quantileDDSketch(0.01, 0.42)(number) FROM numbers_mt(1000000000)
```
It looks accidentally committed.

alexey-milovidov commented on src/AggregateFunctions/Store.h:
```diff
@@ -0,0 +1,267 @@
+    }
+
+private:
+    int chunk_size;
```
It's better to avoid using a signed implementation-defined integer type. If you need a type for sizes, UInt64 is ok; if you need a 32-bit integer type for sizes, UInt32 is ok.

alexey-milovidov reviewed (dismissed):
.

srikanthccv committed:
Address alexey review comments

srikanthccv committed:
Remove unused function isFloat64FieldType

srikanthccv commented:
>Having the files named AggregateFunctions/{Store,Mapping}.h but being specific for DDSketch is no way.

Makes sense. I moved them under sub directory DDSketch. I hope that is fine.

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Throw error on invalid relative accuracy

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Simplify mapping

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
Hi @alexey-milovidov, wanted to ping again ICYMI.

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

alexey-milovidov commented:
I didn't review it, @al13n321 does.

alexey-milovidov commented:
Just as advice - let's compare the performance when there are many small aggregation states.
For example, GROUP BY RegionID, or GROUP BY ClientIP on the ClickBench dataset.

srikanthccv commented:
Thanks, @al13n321 I addressed your review comments. Please take another look.

>Just as advice - let's compare the performance when there are many small aggregation states.
For example, GROUP BY RegionID, or GROUP BY ClientIP on the ClickBench dataset.

Ok, I can do that and share the results/observations.

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
Numbers from running queries on clickbench dataset on my machine.

| Group By           | DDSketch | TDigest | BFloat16 |
| ---------------- | -------- | ------- | -------- |
| RegionID         | 3.4      | 3.5     | 3.4      |
| ClientIP         | 30.5     | 6.6     | 5.3      |
| SearchEngineID   | 2.6      | 2.5     | 2.5      |
| MobilePhoneModel | 1.5      | 1.4     | 1.5      |
| SearchPhrase     | 17       | 10      | 10       |

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
Hi, Happy new years! I am being more persuasive here because there is no good way to work with pre-aggregated histogram sketches with ClickHouse and we need to resort to UDF to achieve that. And that prevents our users from using the ClickHouse Cloud and more (cc @gingerwizard). This is open for quite some time and I would really appreciate another round of reviews or merge if there are no major concers.

al13n321 commented on src/AggregateFunctions/Store.h:
```diff
@@ -0,0 +1,289 @@
+        else
+        {
+            int new_length = getNewLength(new_min_key, new_max_key);
+            bins.resize(new_length, 0.0);
```
> which is reasonable for most practical applications

Ok, but wouldn't it be worth one line of code to cover *all* practical applications, as well as impractical ones? What's the downside of adding that line, other than increasing the code size?

To elaborate on the `relative_accuracy = 1e-6` example: suppose `relative_accuracy = 1e-6`, and the range is [1, 2^64], so there are 100M bins; growing in blocks of 128 means ~780K resizes of average size 50M each, so 39 trillion values get copied, which would take around 5 hours. With exponential growth, around 200M values get copied, which takes more like 100ms. Sure it's a usage error to request so many bins, but having a query silently get stuck ~forever for no apparent reason is still bad user experience.

srikanthccv replied:
Pushed an update in https://github.com/ClickHouse/ClickHouse/pull/56342/commits/42a5f4e350bb5951f740d61925aeb823a75ef452. Let me know if that works.

al13n321 commented:
Ok, plz address the remaining small things and let's get this merged. Sorry about the wait.

(I still feel that this could be 3x less code and 10x more speed without all the OOP, but meh.)

srikanthccv committed:
Address remaining review comments

srikanthccv committed:
Add effective memory usage suggestions

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
Thanks for the review. I pushed updates for new suggestions, please take a look.

> I still feel that this could be 3x less code and 10x more speed without all the OOP

I don't have solid expertise with C++. Few suggestions from the previous review didn't make much difference. I will be happy to make changes if it gives such a significant improvement.

srikanthccv committed:
Fix Stylecheck

srikanthccv committed:
Fix fast test

al13n321 commented on src/AggregateFunctions/DDSketch/Store.h:
```diff
@@ -227,8 +227,14 @@ class DenseStore : public Store
         {
-            int new_length = getNewLength(new_min_key, new_max_key);
-            bins.resize(new_length, 0.0);
+            UInt32 new_length = getNewLength(new_min_key, new_max_key);
+            if (new_length > bins.capacity())
+            {
+                bins.reserve(new_length);
+            }
+            std::size_t old_size = bins.size();
+            bins.resize(new_length);
+            std::fill(bins.begin() + old_size, bins.end(), 0.0);
```
The `if (new_length > bins.capacity())` doesn't change anything, `reserve()` doesn't do anything if requested length is less than current capacity.
The `std::fill()` doesn't do anything because `resize()` already initializes the new elements with default values.

So this simplifies to:
```
            UInt32 new_length = getNewLength(new_min_key, new_max_key);
            bins.reserve(new_length);
            bins.resize(new_length);
```
I.e. adding a reserve() before resize() to disable exponential growth.

Just in case you didn't understand my previous proposal, it was the opposite:
```
bins.resize(new_length);

// (At this point bins.capacity() is usually greater than bins.size(), because resize() grows capacity() exponentially.
//  such that a series of increasing resize()s takes O(final_size) time, even if resizing by one element at a time.
//  But here we do an additional O(n) pass in adjust(), so exponentially growing capacity() doesn't help,
//  we're still O(n^2).
//  So we make size() grow exponentially together with capacity(), by resizing again.)

bins.resize(bins.capacity());
```
(Or maybe I missed some obvious reason why this doesn't work?)

If you understood and decided against it - that's fine.

srikanthccv replied:
No, I misunderstood.

> if (new_length > bins.capacity()) doesn't change anything

I thought your suggestion was to target the case where new length is more than current capacity.

> because resize() already initializes the new elements with default values.

I didn't know about this. I thought the new elements wouldn't contain default values unless explicitly mentioned. Let me make the change and test once.

srikanthccv committed:
Incorporate bins capacity suggestion

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Fix fuzzer

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

al13n321 commented on src/AggregateFunctions/DDSketch/Store.h:
```diff
@@ -228,13 +228,9 @@ class DDSketchDenseStore : public DDSketchStore
             bins.resize(new_length);
-            std::fill(bins.begin() + old_size, bins.end(), 0.0);
+            bins.resize(bins.capacity());
+            new_max_key = new_min_key + int(bins.size()) - 1;
```
Wait, this looks wrong, `new_max_key` should be the max observed key, independent of the array size, no?

srikanthccv replied:
Right, it was incorrect. I reverted it back.

This suggestion was behaving weird for relative accuracy such as 0.0001. I am not sure yet. I will be looking into it.
```
bins.resize(new_length);

// (At this point bins.capacity() is usually greater than bins.size(), because resize() grows capacity() exponentially.
//  such that a series of increasing resize()s takes O(final_size) time, even if resizing by one element at a time.
//  But here we do an additional O(n) pass in adjust(), so exponentially growing capacity() doesn't help,
//  we're still O(n^2).
//  So we make size() grow exponentially together with capacity(), by resizing again.)

bins.resize(bins.capacity());

```

al13n321 commented:
> I don't have solid expertise with C++. Few suggestions from the previous review didn't make much difference. I will be happy to make changes if it gives such a significant improvement.

I would do something like this: https://pastila.nl/?0052513d/409c97e2bb061b060eacf41ae73ceae2#IkYx5MhJoZnbPjohZ4NQNw== (untested), on the off chance you're in the mood for rewriting everything.

al13n321 commented on src/AggregateFunctions/DDSketch/Store.h:
```diff
@@ -228,9 +228,13 @@ class DDSketchDenseStore : public DDSketchStore
             bins.resize(new_length);
-            bins.resize(bins.capacity());
-            new_max_key = new_min_key + int(bins.size()) - 1;
+            std::fill(bins.begin() + old_size, bins.end(), 0.0);
```
Did you mean to put all this back? What's the explanation for the overflow that fuzzer found?

srikanthccv replied:
I am not sure yet. I will spend sometime troubleshoot the cause.

al13n321 replied:
So https://s3.amazonaws.com/clickhouse-test-reports/56342/b55529b5850abc4cc3578b96582c92476c1b558a/fuzzer_astfuzzerubsan/report.html says that this line:
```
int margins = length() - (new_max_key - new_min_key + 1);
```
subtracted 69 from INT32_MIN+68. Presumably the subtraction is `new_max_key - new_min_key`, so new_max_key was INT32_MIN+68, and new_min_key was 69 (:D). So on line 233 int(bins.size()) must've been INT32_MIN, i.e. bins.size() == 2^31.

So I guess it was just very small relative_accuracy and very wide range of values, so max_key - min_key was above 2^30 or so. But that also means that min_key or max_key are probably overflowing all over the place in that test! (Otherwise what are the odds that they came within a factor of 2 of overflowing, but didn't overflow?)

So you need to add overflow checks when calculating keys, throw an exception if it doesn't fit in int32. Or even if it's outside [INT32_MIN/2 + 1, INT32_MAX/2 - 1], so that any two keys can be subtracted without overflow. And do `bins.resize(std::min(bins.capacity(), size_t(INT32_MAX - 1)))`.

srikanthccv commented:
> > I don't have solid expertise with C++. Few suggestions from the previous review didn't make much difference. I will be happy to make changes if it gives such a significant improvement.
> 
> I would do something like this: https://pastila.nl/?0052513d/409c97e2bb061b060eacf41ae73ceae2#IkYx5MhJoZnbPjohZ4NQNw== (untested), on the off chance you're in the mood for rewriting everything.


I will give it a try. If you believe it could improve the performance by multifold I want to give it a try.

al13n321 commented:
> I will give it a try. If you believe it could improve the performance by multifold I want to give it a try.

Cool! Hack a quick prototype first, to check if it's really fast enough to be worth rewriting the rest. (Though it's also less code, so might be worth rewriting even if not much faster, but that's less important and more subjective.)

al13n321 commented:
> > ```
> >    /// (No need to write mapping and scale, they're determined by parameters, which should match
> > ```
> 
> ```
>     ///  between serializer and deserializer. Unless again you need to be compatible with some
>     ///  other system for some reason. No need for changeMapping() for the same reason.
>     ///  Unless I'm missing something?)
> ```
> 
> To give some clarity on this, I was trying to keep the state compatible with the DDSketch proto https://github.com/DataDog/sketches-java/blob/master/src/protobuf/proto/DDSketch.proto. The clickHouse-go client would directly write the rows with the compatible serialized state. This will enable ingesting the pre-aggregated sketches (usually on the web service application) and then computing percentiles at the query time.

Makes sense, please add a comment about it in the code, with the link. Otherwise it just looks like random crazy overengineering tbh :)

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv committed:
Remove inheritance in Mapping, Store and DDSketch

srikanthccv committed:
Add checks for bin resizing

srikanthccv committed:
Add note about the binary compatible implementation

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
@al13n321 I tried the suggestion in a separate branch and it improved the performance by ~20%. And then I revisited your comment here about the virtual functions https://github.com/ClickHouse/ClickHouse/pull/56342#discussion_r1406777094. The last time when I added `final` it didn't make a difference but I removed the inheritance and virtual functions and that change also improved the performance, i.e. ~20%. I pushed the updated changes, please take a look again.

srikanthccv committed:
Merge remote-tracking branch 'upstream/master' into ddsketch-quantile

srikanthccv commented:
I started looking into small object optimization. Is it okay if that is addressed in the follow-up PR? This has gone through many changes at this point and has been reviewed. Since it will be an enhancement that doesn't break anything, I think it's better to address it in a follow-up PR without adding more new changes to this already reviewed PR. Let me know what do you think?

alexey-milovidov commented:
@srikanthccv, totally ok.

al13n321 commented:
Looks mergable, except for the fuzzer failure from before: https://s3.amazonaws.com/clickhouse-test-reports/56342/b55529b5850abc4cc3578b96582c92476c1b558a/fuzzer_astfuzzerubsan/report.html - did you explain or fix it? It didn't happen this time, but I don't see anything that looks like a fix, so it's probably not fixed and will come back.

Maybe try to reproduce it by running the last few queries with substring "DDSketch" from `fuzzer.log` and `server.log.zst ` (from the link above). If that doesn't work, just add an overflow check in `getNewLength()`, and we'll hope that my explanation in https://github.com/ClickHouse/ClickHouse/pull/56342#discussion_r1446648028 is correct.

srikanthccv commented:
hmm, I posted a comment before resolved but it seems to have been not delivered.

@al13n321 The [commit](https://github.com/ClickHouse/ClickHouse/pull/56342/commits/229a1e48cc02868a586a2aa6606e8bfa8c9e49a0) I made which caused the overflow was an incorrect change and shouldn't have been committed.

Here is what happened. The `new_length` returned occasionally would be smaller than the length of the current bins after the suggestion `bins.resize(bins.capacity());` is incorporated - the size of the current bins is enough to adjust the layout. Originally there was no check before resizing and as a result, some previous counts got wiped out. I added the check in this commit [`d831792` (#56342)](https://github.com/ClickHouse/ClickHouse/pull/56342/commits/d831792061b4ef545e6400c995fa73f32868476c). The overflow shouldn't happen the keys are going to be well within the range. We do not allow relative accuracy no less than 1e-6. So the max possible value is `1.79769e+308` and it's the corresponding key is `354891355` so we are fine. Let me know if this answers your question.

srikanthccv commented:
@al13n321 @alexey-milovidov @bharatnc Thanks for the reviews and help. I will send a follow-up PR for small object optimization after testing against the ClickBench data.
